apps/backend/app/services/model_training/feature_engineer.py

```
@@ -664,4 +664,43 @@ def _apply_generic(
             # Default to multiplication for two columns
             df[name] = df[col1] * df[col2]
 
-        return df
\ No newline at end of file
+        return df
+
+    async def apply_stored_feature(
+        self,
+        df: pd.DataFrame,
+        feature: 'StoredFeature'
```

Ruff (0.14.10)

Undefined name `StoredFeature`

(F821)